### ISSUE TITLE
Insert Media. Button active too soon resulting in no image.

### ISSUE THREAD

**Issue by MattyBalaam** [silverstripe/silverstripe-cms#1166]
When uploading an image the 'Insert' button becomes active before the image is properly attached. If this is clicked nothing gets inserted into the editor.
![insert-media](https://cloud.githubusercontent.com/assets/1246923/6186879/b772d382-b377-11e4-865d-0315219fdfcb.png)

**Comment by gordonbanderson** [comment]
Whilst I have not solved this I am testing locally and observing as follows:

1) Limit upload speed to GPRS using Chrome's dev tools, the lack of network speed gives time to click the Insert button a few times.  See https://developer.chrome.com/devtools/docs/device-mode#network-conditions

2) Adding console trace and recreating the bug by clicking on the Insert button feverishly prior to an image uploading I note:

* The area on screen containing the list of files does not appear visually.
* The Insert button has been enabled.
* An attempt to insert the images is made, but the HTML for the image is incomplete, meaning that a blank paragraph is inserted into the HTML instead of an image.  By 'HTML for an image is incomplete' I mean that the HTML to being inserted when the bug appears is this:

```HTML
<div class="ss-htmleditorfield-file loading"></div>
```

instead of likes of this:

```HTML

<div class="ss-uploadfield-item image ss-htmleditorfield-file template-upload" data-id="162" data-url="assets/Uploads/screenshot741.png">
	
	
	<div class="ss-uploadfield-item-preview">
		
			<span><img src="/assets/Uploads/_resampled/SetRatioSize3040-CMSThumbnail-screenshot741.png" alt="screenshot741"></span>
		
	</div>
	
	<div class="ss-uploadfield-item-info">
		<label class="ss-uploadfield-item-name">
			<span class="name" title="screenshot741.png">
				screenshot741.png
			</span>
			
			<div class="ss-uploadfield-item-status ui-state-success-text" title="Dimensions">
				1280 x 800 (px)
			</div>
			

			<div class="clear"><!-- --></div> 
		</label>
		<div class="ss-uploadfield-item-actions">	
			<button data-icon="deleteLight" class="ss-uploadfield-item-cancel ss-uploadfield-item-remove ss-ui-button" title="Remove">
				Remove
			</button>
			
			<div class="ss-uploadfield-item-edit edit">
				<button class="ss-uploadfield-item-edit ss-ui-button" title="Edit this file" data-icon="pencil">
					Edit
					<span class="toggle-details">
						<span class="toggle-details-icon"></span>
					</span>
				</button>
			</div>
		</div>
		
		<div class="details ss-uploadfield-item-editform" style="height: 0px;">
			<fieldset>
				
					<div class="CompositeField composite cms-file-info nolabel ">
	
	
	
		
			<div class="CompositeField composite cms-file-info-preview nolabel ">
	
	
	
		
			<img id="thumbnailImage" class="thumbnail-preview" src="assets/Uploads/screenshot741.png?r=75640" alt="screenshot741.png">

		
	

	
</div>

		
	
		
			<div class="CompositeField composite cms-file-info-data nolabel ">
	
	
	
		
			<div class="CompositeField composite nolabel ">
	
	
	
		
			<div id="FileType" class="field readonly">
	<label class="left" for="FileType">File type</label>
	<div class="middleColumn">
		
	<span id="FileType" class="readonly">
		PNG image - good general-purpose format
	</span>


	</div>
	
	
	
</div>

		
	
		
			<div id="Size" class="field readonly">
	<label class="left" for="Size">File size</label>
	<div class="middleColumn">
		
	<span id="Size" class="readonly">
		575 KB
	</span>


	</div>
	
	
	
</div>

		
	
		
			<div id="ClickableURL" class="field readonly">
	<label class="left" for="ClickableURL">URL</label>
	<div class="middleColumn">
		
	<span id="ClickableURL" class="readonly">
		<a href="/assets/Uploads/screenshot741.png" title="/assets/Uploads/screenshot741.png" target="_blank" class="file-url">assets/Uploads/screenshot741.png</a>
	</span>


	</div>
	
	
	
</div>

		
	
		
			<div id="Created" class="field date_disabled readonly">
	<label class="left" for="Created">First uploaded</label>
	<div class="middleColumn">
		<span class="readonly" id="Created">Jul 19, 2015 (today)</span>
	</div>
	
	
	
</div>

		
	
		
			<div id="LastEdited" class="field date_disabled readonly">
	<label class="left" for="LastEdited">Last changed</label>
	<div class="middleColumn">
		<span class="readonly" id="LastEdited">Jul 19, 2015 (today)</span>
	</div>
	
	
	
</div>

		
	

	
</div>

		
	

	
</div>

		
	

	
</div>

				
					<div id="AltText" class="field text">
	<label class="left" for="AltText">Alternative text (alt)</label>
	<div class="middleColumn">
		<input type="text" name="AltText" value="screenshot741" class="text" id="AltText" maxlength="80" size="30">
	</div>
	
	
	<span class="description">Shown to screen readers or if image can not be displayed</span>
</div>

				
					<div id="Title" class="field text">
	<label class="left" for="Title">Title text (tooltip)</label>
	<div class="middleColumn">
		<input type="text" name="Title" class="text" id="Title">
	</div>
	
	
	<span class="description">For additional information about the image</span>
</div>

				
					<div id="CaptionText" class="field text">
	<label class="left" for="CaptionText">Caption text</label>
	<div class="middleColumn">
		<input type="text" name="CaptionText" class="text" id="CaptionText">
	</div>
	
	
	
</div>

				
					<div id="CSSClass" class="field dropdown last">
	<label class="left" for="CSSClass">Alignment / style</label>
	<div class="middleColumn">
		<select name="CSSClass" class="dropdown last has-chzn chzn-done" id="CSSClass" style="display: none;">

	<option value="tinymce leftAlone">On the left, on its own.</option>

	<option value="tinymce center">Centered, on its own.</option>

	<option value="tinymce left">On the left, with text wrapping around.</option>

	<option value="tinymce right">On the right, with text wrapping around.</option>

</select><div id="CSSClass_chzn" class="chzn-container chzn-container-single chzn-with-rise chzn-container-single-nosearch" style="width: 450px;"><a href="javascript:void(0)" class="chzn-single" tabindex="-1"><span>On the left, on its own.</span><div><b></b></div></a><div class="chzn-drop" style="left: -9000px; width: 414px; top: -40px;"><div class="chzn-search"><input type="text" autocomplete="off" style="width: 379px;"></div><ul class="chzn-results"><li id="CSSClass_chzn_o_0" class="active-result result-selected" style="">On the left, on its own.</li><li id="CSSClass_chzn_o_1" class="active-result" style="">Centered, on its own.</li><li id="CSSClass_chzn_o_2" class="active-result" style="">On the left, with text wrapping around.</li><li id="CSSClass_chzn_o_3" class="active-result" style="">On the right, with text wrapping around.</li></ul></div></div>

	</div>
	
	
	
</div>

				
					<div class="field fieldgroup dimensions last">
	<label class="left">Dimensions</label>
	
	<div class="middleColumn fieldgroup">
		
			<div class="fieldgroup-field first odd">
				<div class="fieldholder-small">
	<label class="fieldholder-small-label" for="Width">Width</label>
	<input type="text" name="Width" value="1000" class="text" id="Width" maxlength="5" size="5">
	
</div>
			</div>
		
			<div class="fieldgroup-field last even">
				<div class="fieldholder-small">
	<label class="fieldholder-small-label" for="Height"> x Height</label>
	<input type="text" name="Height" value="625" class="text" id="Height" maxlength="5" size="5">
	
</div>
			</div>
		
	</div>
	
	
	
</div>

				
					<input type="hidden" name="URL" value="assets/Uploads/screenshot741.png" class="hidden nolabel" id="URL">

				
			</fieldset>
		</div>
	</div>
</div>
```

In Tiny MCE this is converted to this HTML upon a succesful insertion:
```HTML
<p><img class="tinymce leftAlone" title="" src="assets/Uploads/screenshot741.png" alt="screenshot741" width="1000" height="625" /></p>
```

**Comment by gordonbanderson** [comment]
I think something sneaky is going on HtmlEditorField.js around the event onfileuploadstop , possibly the file uploader triggering the event too early?  Just hypothesizing though.  Note I am testing against 3.1.13 BTW

**Comment by gordonbanderson** [comment]
One possible kludgy fix is to  check that the HTML for the uploaded file is correct during the onsubmit method after the Insert button is clicked.  Method as below.  It would involve checking for .ss-htmleditorfield-file.loading and if that exists, not hiding the form.

Still feels and icky route to a solution though and does not address the root cause of the problem though.


```JavaScript
onsubmit: function() {
				console.log('----------------');
				console.log('Submission of insert media');
				this.modifySelection(function(ed){
					this.find('.ss-htmleditorfield-file').each(function() {
						console.log("Inserting", ed, this);
						$(this).insertHTML(ed);
					});

					console.log('Repaint');

					ed.repaint();
				});

				console.log('Close dialog');

				this.getDialog().close();
				return false;
			},
```

**Comment by gordonbanderson** [comment]
Think I have finally got to the bottom of this.

The Insert button is enabled in the method redraw, https://github.com/silverstripe/silverstripe-framework/blob/3.1.3/javascript/HtmlEditorField.js#L848

https://github.com/silverstripe/silverstripe-framework/blob/3.1.3/javascript/HtmlEditorField.js#L861 enables or disables the button.  The following line enables or disables the button visually.  The toggle depends on a variable called hasItems which is defined as

```JavaScript
hasItems = Boolean(this.find('.ss-htmleditorfield-file').length)
```

When the jquery file-fileupload UI triggers the event 'fileuploadstop' (https://github.com/silverstripe/silverstripe-framework/blob/3.1.3/javascript/HtmlEditorField.js#L935) each of the uploaded files are iterated and a method called form.showFileView() is called (line 951).  This appears to be triggered asynchronously due to the use of entwine (console logging certainly suggests so).  The form is then redrawn, and because we now have at least one file uploaded to a DOM element with class '.ss-htmleditorfield-file', the insert button is enabled.

*However* there is no guarantee that the element containing the uploaded file has been redrawn yet - another AJAX called is made at https://github.com/silverstripe/silverstripe-framework/blob/3.1.3/javascript/HtmlEditorField.js#L896 to render the HTML with appropriate hooks to be inserted into TinyMCE correctly.  And of course a large uploaded image may be slow in generating a thumbnail.

So the two lines of execution here are:
CORRECT) File is uploaded, fileuploadstopevent is triggered enabling redraw of the form and enabling  the Insert button. The upload field is re-rendered with appropriate hooks to insert image into TinyMCE, upload works as expected when the Insert button is clicked.
INCORRECT) File is uploaded, fileuploadstopevent is triggered enabling redraw of the form and enabling  the Insert button.  The Insert button is clicked before the upload field(s) have been re-rendered with hooks for Tiny MCE.  An attempt to insert is made by clicking the Insert button, but due to lack of appropriate information nothing is inserted.  However the situation is recoverable if one waits until the Upload field has re-rendering itself for TinyMCE.

So in short it appears the Insert button should only be enabled after the UploadField DOM Elements have been re-rendered for use within TinyMCE.  It is currently enabled too early leading to this bug.

**Comment by gordonbanderson** [comment]
I've added my modified version of HtmlFieldEditor.js (from SS 3.13) here, the addition being console logging, along with trace of a normal successful and then a failed insertion https://gist.github.com/gordonbanderson/1349bce77cc51177d829

The last one is the most crucial:

Click on loading a file and either upload or overwrite (as I was when testing).  Note that Chrome can throttle browser speeds which is a good way to test this.
```
Update from editor
HtmlEditorField.js?m=1437392823:962 Redraw
HtmlEditorField.js?m=1437392823:970 Has items false
HtmlEditorField.js?m=1437392823:973 Number of files 0
HtmlEditorField.js?m=1437392823:979 Enable/Disable insert button ui state disabled true
HtmlEditorField.js?m=1437392823:991 cleaning UI allegedly
HtmlEditorField.js?m=1437392823:962 Redraw
HtmlEditorField.js?m=1437392823:970 Has items false
HtmlEditorField.js?m=1437392823:973 Number of files 0
HtmlEditorField.js?m=1437392823:979 Enable/Disable insert button ui state disabled true
HtmlEditorField.js?m=1437392823:991 cleaning UI allegedly
HtmlEditorField.js?m=1437392823:1068 file upload done
HtmlEditorField.js?m=1437392823:1077 File upload stop
```
File has now been uploaded but due to throttled browser speed, the AJAX called to re-render the upload widget will take time.  Trigger the AJAX call on the uploaded file element, and then redraw the form.  This redraw enables the Insert button incorrectly.
```
HtmlEditorField.js?m=1437392823:1091 Uploaded file undefined
HtmlEditorField.js?m=1437392823:1015 show file view
HtmlEditorField.js?m=1437392823:1100 About to redraw form
HtmlEditorField.js?m=1437392823:962 Redraw
HtmlEditorField.js?m=1437392823:970 Has items true
HtmlEditorField.js?m=1437392823:973 Number of files 1
HtmlEditorField.js?m=1437392823:979 Enable/Disable insert button ui state disabled false
HtmlEditorField.js?m=1437392823:991 cleaning UI allegedly
```
Now insert the media.  Note this is prior to the uploaded field elements being re-rendered for Tiny MCE.
```
HtmlEditorField.js?m=1437392823:929 ----------------
HtmlEditorField.js?m=1437392823:930 Submission of insert media
HtmlEditorField.js?m=1437392823:933 Inserting Object {instance: k.c…e.Editor} <div class=​"ss-htmleditorfield-file loading">​</div>​
HtmlEditorField.js?m=1437392823:1193 **** INSERT HTML ENTWINE ****
HtmlEditorField.js?m=1437392823:1182 **** GET HTML ENTWINE ****
HtmlEditorField.js?m=1437392823:1194 <a></a>
HtmlEditorField.js?m=1437392823:1182 **** GET HTML ENTWINE ****
HtmlEditorField.js?m=1437392823:937 Repaint
HtmlEditorField.js?m=1437392823:942 Close dialog
HtmlEditorField.js?m=1437392823:1002 Reset fields
HtmlEditorField.js?m=1437392823:962 Redraw
HtmlEditorField.js?m=1437392823:970 Has items false
HtmlEditorField.js?m=1437392823:973 Number of files 0
HtmlEditorField.js?m=1437392823:979 Enable/Disable insert button ui state disabled true
HtmlEditorField.js?m=1437392823:991 cleaning UI allegedly
```

Nothing is inserted into TinyMCE. Only now is an attempt made to redraw the successfully uploaded fields, but they no longer exist having being removed in the previous action.
```
jquery.js?m=1432796351:8206 Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user's experience. For more help, check http://xhr.spec.whatwg.org/.
HtmlEditorField.js?m=1437392823:1027 T1 AJAX SUCCESS REDRAW UPLOAD FIELD
HtmlEditorField.js?m=1437392823:1029 T2 AJAX SUCCESS REDRAW UPLOAD FIELD
HtmlEditorField.js?m=1437392823:1031 T3 AJAX SUCCESS REDRAW UPLOAD FIELD
HtmlEditorField.js?m=1437392823:962 Redraw
HtmlEditorField.js?m=1437392823:970 Has items false
HtmlEditorField.js?m=1437392823:973 Number of files 0
HtmlEditorField.js?m=1437392823:979 Enable/Disable insert button ui state disabled true
HtmlEditorField.js?m=1437392823:991 cleaning UI allegedly
HtmlEditorField.js?m=1437392823:1033 T4 AJAX SUCCESS REDRAW UPLOAD FIELD
HtmlEditorField.js?m=1437392823:1035 T5 AJAX SUCCESS REDRAW UPLOAD FIELD
```
Upload in progress:
![uploading001](https://cloud.githubusercontent.com/assets/7060/8776492/cb9db5fa-2f1b-11e5-8f60-ec54ceb8a0b8.png)

The 'at risk' period when Insert button has been enabled but insertion will fail
![uploading002](https://cloud.githubusercontent.com/assets/7060/8776499/e36340f6-2f1b-11e5-82f7-f69fda580134.png)

Clicking now will work
![uploading003](https://cloud.githubusercontent.com/assets/7060/8776508/ed5c7816-2f1b-11e5-8249-98d7479df92c.png)

**Comment by purplespider** [comment]
This is still an issue in 4.0.0-alpha3, and is something that even trips me up from time to time despite knowing about it.

**Comment by MattyBalaam** [comment]
This trips up the content authors on the sites I manage. Seems like a huge lack of attention to detail for me in an otherwise excellent platform.

**Comment by MattyBalaam** [comment]
OK, I did some digging and there is a variable in HtmlEditorField.js in redraw called hasItems. If the condition it checks against is changed to `Boolean(this.find('.ss-htmleditorfield-file.template-upload').length)` then the button will stay disabled until images are ready, in my very limited testing.

**Comment by dhensby** [comment]
@MattyBalaam ah, so it really needs to check for *completed* items?

**Comment by lukereative** [comment]
Closing this as the insert media dialog has completely changed in SS4.x and considering that 3.x is in security only support this issue is not going to be fixed for 3.x
